File: src/primitives/text.rs
//! Text Primitive - Display text with styling and wrapping.
//!
//! A pure display component for text content. Cannot have children.
//!
//! # Reactivity
//!
//! Content can be a static string, signal, or getter. When the content
//! source changes, the display updates automatically.
//!
//! # Example
//!
//! ```ignore
//! use spark_tui::primitives::{text, TextProps};
//! use spark_signals::signal;
//!
//! // Static text
//! text(TextProps {
//!     content: "Hello, World!".to_string().into(),
//!     ..Default::default()
//! });
//!
//! // Reactive text
//! let count = signal(0);
//! let count_clone = count.clone();
//! text(TextProps {
//!     content: PropValue::Getter(Rc::new(move || format!("Count: {}", count_clone.get()))),
//!     attrs: Some(Attr::BOLD.into()),
//!     ..Default::default()
//! });
//!
//! // Update count - text updates automatically
//! count.set(42);
//! ```

use crate::engine::{
    allocate_index, release_index, create_flex_node,
    get_current_parent_index,
};
use crate::engine::arrays::{core, visual, text as text_arrays};
use crate::state::mouse;
use crate::types::ComponentType;
use super::types::{TextProps, PropValue, Cleanup};

// =============================================================================
// Text Component
// =============================================================================

/// Create a text display component.
///
/// Text is used to display strings with optional styling (bold, italic, etc.),
/// alignment, and wrapping behavior.
///
/// # Properties
///
/// - `content` - The text to display (required)
/// - `attrs` - Text attributes like bold, italic, underline
/// - `align` - Text alignment: left, center, right
/// - `wrap` - Wrap mode: wrap, nowrap, truncate
///
/// Returns a cleanup function that releases resources when called.
pub fn text(props: TextProps) -> Cleanup {
    // 1. ALLOCATE INDEX
    let index = allocate_index(props.id.as_deref());

    // 2. CREATE FLEXNODE - Even text needs layout properties
    let flex_node = create_flex_node(index);

    // 3. CORE SETUP - Type, parent
    core::set_component_type(index, ComponentType::Text);
    if let Some(parent) = get_current_parent_index() {
        core::set_parent_index(index, Some(parent));
    }

    // 4. BIND VISIBILITY
    if let Some(visible) = props.visible {
        match visible {
            PropValue::Static(v) => core::set_visible(index, v),
            PropValue::Signal(s) => core::set_visible_signal(index, s),
            PropValue::Getter(g) => core::set_visible_getter(index, move || g()),
        }
    }

    // 5. BIND TEXT CONTENT
    match props.content {
        PropValue::Static(v) => text_arrays::set_text_content(index, v),
        PropValue::Signal(s) => text_arrays::set_text_content_signal(index, s),
        PropValue::Getter(g) => text_arrays::set_text_content_getter(index, move || g()),
    }

    // 6. BIND TEXT STYLING
    if let Some(attrs) = props.attrs {
        match attrs {
            PropValue::Static(v) => text_arrays::set_text_attrs(index, v),
            PropValue::Signal(s) => text_arrays::set_text_attrs_getter(index, move || s.get()),
            PropValue::Getter(g) => text_arrays::set_text_attrs_getter(index, move || g()),
        }
    }

    if let Some(align) = props.align {
        match align {
            PropValue::Static(v) => text_arrays::set_text_align(index, v),
            PropValue::Signal(s) => text_arrays::set_text_align_getter(index, move || s.get()),
            PropValue::Getter(g) => text_arrays::set_text_align_getter(index, move || g()),
        }
    }

    if let Some(wrap) = props.wrap {
        match wrap {
            PropValue::Static(v) => text_arrays::set_text_wrap(index, v),
            PropValue::Signal(s) => text_arrays::set_text_wrap_getter(index, move || s.get()),
            PropValue::Getter(g) => text_arrays::set_text_wrap_getter(index, move || g()),
        }
    }

    // 7. BIND FLEXNODE SLOTS - Layout properties

    // Item properties (text is always a flex item, never a container)
    if let Some(grow) = props.grow {
        match grow {
            PropValue::Static(v) => flex_node.flex_grow.set_value(v),
            PropValue::Signal(s) => flex_node.flex_grow.set_signal(s),
            PropValue::Getter(g) => flex_node.flex_grow.set_signal_readonly(move || g()),
        }
    }
    if let Some(shrink) = props.shrink {
        match shrink {
            PropValue::Static(v) => flex_node.flex_shrink.set_value(v),
            PropValue::Signal(s) => flex_node.flex_shrink.set_signal(s),
            PropValue::Getter(g) => flex_node.flex_shrink.set_signal_readonly(move || g()),
        }
    }
    if let Some(basis) = props.flex_basis {
        match basis {
            PropValue::Static(v) => flex_node.flex_basis.set_value(v),
            PropValue::Signal(s) => flex_node.flex_basis.set_signal(s),
            PropValue::Getter(g) => flex_node.flex_basis.set_signal_readonly(move || g()),
        }
    }
    if let Some(align_self) = props.align_self {
        match align_self {
            PropValue::Static(v) => flex_node.align_self.set_value(v),
            PropValue::Signal(s) => flex_node.align_self.set_signal(s),
            PropValue::Getter(g) => flex_node.align_self.set_signal_readonly(move || g()),
        }
    }

    // Dimensions
    if let Some(w) = props.width {
        match w {
            PropValue::Static(v) => flex_node.width.set_value(v),
            PropValue::Signal(s) => flex_node.width.set_signal(s),
            PropValue::Getter(g) => flex_node.width.set_signal_readonly(move || g()),
        }
    }
    if let Some(h) = props.height {
        match h {
            PropValue::Static(v) => flex_node.height.set_value(v),
            PropValue::Signal(s) => flex_node.height.set_signal(s),
            PropValue::Getter(g) => flex_node.height.set_signal_readonly(move || g()),
        }
    }
    if let Some(min_w) = props.min_width {
        match min_w {
            PropValue::Static(v) => flex_node.min_width.set_value(v),
            PropValue::Signal(s) => flex_node.min_width.set_signal(s),
            PropValue::Getter(g) => flex_node.min_width.set_signal_readonly(move || g()),
        }
    }
    if let Some(max_w) = props.max_width {
        match max_w {
            PropValue::Static(v) => flex_node.max_width.set_value(v),
            PropValue::Signal(s) => flex_node.max_width.set_signal(s),
            PropValue::Getter(g) => flex_node.max_width.set_signal_readonly(move || g()),
        }
    }
    if let Some(min_h) = props.min_height {
        match min_h {
            PropValue::Static(v) => flex_node.min_height.set_value(v),
            PropValue::Signal(s) => flex_node.min_height.set_signal(s),
            PropValue::Getter(g) => flex_node.min_height.set_signal_readonly(move || g()),
        }
    }
    if let Some(max_h) = props.max_height {
        match max_h {
            PropValue::Static(v) => flex_node.max_height.set_value(v),
            PropValue::Signal(s) => flex_node.max_height.set_signal(s),
            PropValue::Getter(g) => flex_node.max_height.set_signal_readonly(move || g()),
        }
    }

    // Spacing - Padding
    if let Some(ref p) = props.padding {
        if props.padding_top.is_none() {
            match p.clone() {
                PropValue::Static(v) => flex_node.padding_top.set_value(v),
                PropValue::Signal(s) => flex_node.padding_top.set_signal(s),
                PropValue::Getter(g) => flex_node.padding_top.set_signal_readonly(move || g()),
            }
        }
        if props.padding_right.is_none() {
            match p.clone() {
                PropValue::Static(v) => flex_node.padding_right.set_value(v),
                PropValue::Signal(s) => flex_node.padding_right.set_signal(s),
                PropValue::Getter(g) => flex_node.padding_right.set_signal_readonly(move || g()),
            }
        }
        if props.padding_bottom.is_none() {
            match p.clone() {
                PropValue::Static(v) => flex_node.padding_bottom.set_value(v),
                PropValue::Signal(s) => flex_node.padding_bottom.set_signal(s),
                PropValue::Getter(g) => flex_node.padding_bottom.set_signal_readonly(move || g()),
            }
        }
        if props.padding_left.is_none() {
            match p.clone() {
                PropValue::Static(v) => flex_node.padding_left.set_value(v),
                PropValue::Signal(s) => flex_node.padding_left.set_signal(s),
                PropValue::Getter(g) => flex_node.padding_left.set_signal_readonly(move || g()),
            }
        }
    }
    if let Some(pt) = props.padding_top {
        match pt {
            PropValue::Static(v) => flex_node.padding_top.set_value(v),
            PropValue::Signal(s) => flex_node.padding_top.set_signal(s),
            PropValue::Getter(g) => flex_node.padding_top.set_signal_readonly(move || g()),
        }
    }
    if let Some(pr) = props.padding_right {
        match pr {
            PropValue::Static(v) => flex_node.padding_right.set_value(v),
            PropValue::Signal(s) => flex_node.padding_right.set_signal(s),
            PropValue::Getter(g) => flex_node.padding_right.set_signal_readonly(move || g()),
        }
    }
    if let Some(pb) = props.padding_bottom {
        match pb {
            PropValue::Static(v) => flex_node.padding_bottom.set_value(v),
            PropValue::Signal(s) => flex_node.padding_bottom.set_signal(s),
            PropValue::Getter(g) => flex_node.padding_bottom.set_signal_readonly(move || g()),
        }
    }
    if let Some(pl) = props.padding_left {
        match pl {
            PropValue::Static(v) => flex_node.padding_left.set_value(v),
            PropValue::Signal(s) => flex_node.padding_left.set_signal(s),
            PropValue::Getter(g) => flex_node.padding_left.set_signal_readonly(move || g()),
        }
    }

    // 8. BIND VISUAL ARRAYS

    // Foreground color
    if let Some(fg) = props.fg {
        match fg {
            PropValue::Static(v) => visual::set_fg_color(index, v),
            PropValue::Signal(s) => visual::set_fg_color_getter(index, move || s.get()),
            PropValue::Getter(g) => visual::set_fg_color_getter(index, move || g()),
        }
    }

    // Background color
    if let Some(bg) = props.bg {
        match bg {
            PropValue::Static(v) => visual::set_bg_color(index, v),
            PropValue::Signal(s) => visual::set_bg_color_getter(index, move || s.get()),
            PropValue::Getter(g) => visual::set_bg_color_getter(index, move || g()),
        }
    }

    // Opacity
    if let Some(opacity) = props.opacity {
        match opacity {
            PropValue::Static(v) => visual::set_opacity(index, v),
            PropValue::Signal(s) => visual::set_opacity(index, s.get()),
            PropValue::Getter(g) => visual::set_opacity(index, g()),
        }
    }

    // 9. REGISTER MOUSE HANDLER (if on_click provided)
    let mut mouse_cleanup: Option<Box<dyn FnOnce()>> = None;

    if let Some(on_click) = props.on_click.clone() {
        let handlers = mouse::MouseHandlers {
            on_mouse_down: None,
            on_mouse_up: None,
            on_click: Some(on_click),
            on_mouse_enter: None,
            on_mouse_leave: None,
            on_scroll: None,
        };
        let cleanup_fn = mouse::on_component(index, handlers);
        mouse_cleanup = Some(Box::new(cleanup_fn));
    }

    // 10. RETURN CLEANUP
    Box::new(move || {
        // Clean up mouse handler
        if let Some(cleanup) = mouse_cleanup {
            cleanup();
        }
        // Clean up component state in mouse module
        mouse::cleanup_index(index);
        // Release index
        release_index(index);
    })
}

// =============================================================================
// Tests
// =============================================================================

#[cfg(test)]
mod tests {
    use super::*;
    use crate::engine::reset_registry;
    use crate::types::{Attr, TextAlign, TextWrap};
    use spark_signals::signal;

    fn setup() {
        reset_registry();
    }

    #[test]
    fn test_text_creation() {
        setup();

        let cleanup = text(TextProps {
            content: PropValue::Static("Hello, World!".to_string()),
            ..Default::default()
        });

        // Check component was created
        assert_eq!(core::get_component_type(0), ComponentType::Text);
        assert_eq!(text_arrays::get_text_content(0), "Hello, World!");

        // Cleanup
        cleanup();
        assert_eq!(core::get_component_type(0), ComponentType::None);
    }

    #[test]
    fn test_text_reactive_content() {
        setup();

        let content = signal("Initial".to_string());
        let content_for_text = content.clone();

        let _cleanup = text(TextProps {
            content: PropValue::Signal(content_for_text),
            ..Default::default()
        });

        assert_eq!(text_arrays::get_text_content(0), "Initial");

        // Update signal - text should update
        content.set("Updated".to_string());
        assert_eq!(text_arrays::get_text_content(0), "Updated");
    }

    #[test]
    fn test_text_with_attrs() {
        setup();

        let _cleanup = text(TextProps {
            content: PropValue::Static("Bold Text".to_string()),
            attrs: Some(PropValue::Static(Attr::BOLD | Attr::ITALIC)),
            ..Default::default()
        });

        assert_eq!(text_arrays::get_text_attrs(0), Attr::BOLD | Attr::ITALIC);
    }

    #[test]
    fn test_text_alignment() {
        setup();

        let _cleanup = text(TextProps {
            content: PropValue::Static("Centered".to_string()),
            align: Some(PropValue::Static(TextAlign::Center)),
            wrap: Some(PropValue::Static(TextWrap::NoWrap)),
            ..Default::default()
        });

        assert_eq!(text_arrays::get_text_align(0), TextAlign::Center);
        assert_eq!(text_arrays::get_text_wrap(0), TextWrap::NoWrap);
    }

    #[test]
    fn test_text_in_box() {
        setup();

        use super::super::{box_primitive, BoxProps};

        let _cleanup = box_primitive(BoxProps {
            children: Some(Box::new(|| {
                text(TextProps {
                    content: PropValue::Static("Child Text".to_string()),
                    ..Default::default()
                });
            })),
            ..Default::default()
        });

        // Parent box at index 0, text at index 1
        assert_eq!(core::get_component_type(0), ComponentType::Box);
        assert_eq!(core::get_component_type(1), ComponentType::Text);
        assert_eq!(core::get_parent_index(1), Some(0));
        assert_eq!(text_arrays::get_text_content(1), "Child Text");
    }
}

File: ts/primitives/text.ts
/**
 * TUI Framework - Text Primitive (v3 Buffer)
 *
 * Display text with styling, alignment, and wrapping.
 * Uses v3 SharedBuffer (1024-byte stride) with Grid support.
 *
 * REACTIVITY: All props flow through repeat() → SharedSlotBuffer → SharedArrayBuffer.
 * Text content uses a SINGLE repeater — the readFn encodes UTF-8 bytes into the
 * text pool and marks the node dirty. Rust handles text measurement via Taffy.
 *
 * Usage:
 * ```ts
 * const msg = signal('Hello')
 * text({ content: msg })
 * msg.value = 'Updated!'  // repeat() fires inline → pool write → Rust wakes
 * ```
 */

import { repeat } from '@rlabs-inc/signals'
import { ComponentType } from '../types'
import type { RGBA } from '../types'
import {
  allocateIndex,
  releaseIndex,
  getCurrentParentIndex,
  registerParent,
} from '../engine/registry'
import {
  pushCurrentComponent,
  popCurrentComponent,
  runMountCallbacks,
} from '../engine/lifecycle'
import { cleanupIndex as cleanupKeyboardListeners } from '../state/keyboard'
import { onComponent as onMouseComponent } from '../state/mouse'
import { getVariantStyle } from '../state/theme'
import { getActiveScope } from './scope'
import { getArrays, getBuffer } from '../bridge'
import {
  packColor,
  setText,
  getU32,
  N_TEXT_OFFSET,
  DIRTY_TEXT,
  markDirty,
  type SharedBuffer,
} from '../bridge/shared-buffer'
import type { TextProps, Cleanup, GridLine } from './types'

// =============================================================================
// CONVERSION HELPERS
// =============================================================================

/** Dimension → Taffy float: NaN = auto, negative = percentage, positive = pixels */
function toDim(dim: number | string | undefined | null): number {
  if (dim === undefined || dim === null || dim === 0) return NaN
  if (typeof dim === 'string') {
    if (dim.endsWith('%')) return -parseFloat(dim) // '100%' → -100.0
    return parseFloat(dim) || NaN
  }
  return dim
}

function unwrap<T>(prop: T | (() => T) | { readonly value: T }): T {
  if (typeof prop === 'function') return (prop as () => T)()
  if (prop !== null && typeof prop === 'object' && 'value' in prop) return (prop as { value: T }).value
  return prop
}

function isReactive(prop: unknown): boolean {
  return typeof prop === 'function' || (prop !== null && typeof prop === 'object' && 'value' in (prop as any))
}

function toPackedColor(c: RGBA | number | null | undefined): number {
  if (c === null || c === undefined) return 0
  if (typeof c === 'number') return c
  return packColor(c.r, c.g, c.b, c.a ?? 255)
}

function dimInput(prop: TextProps['width']): number | (() => number) {
  if (prop === undefined) return NaN
  if (typeof prop === 'number' || typeof prop === 'string') return toDim(prop)
  return () => toDim(unwrap(prop))
}

function enumInput(prop: unknown, converter: (v: any) => number): number | (() => number) {
  if (prop === undefined) return converter(undefined)
  if (typeof prop === 'string') return converter(prop)
  if (isReactive(prop)) return () => converter(unwrap(prop))
  return converter(prop as string)
}

function colorInput(prop: TextProps['fg']): number | (() => number) {
  if (prop === undefined) return 0
  if (!isReactive(prop)) return toPackedColor(prop as RGBA | number | null)
  return () => toPackedColor(unwrap(prop as any))
}

function numInput(prop: unknown, defaultVal = 0): number | (() => number) | { readonly value: number } {
  if (prop === undefined) return defaultVal
  return prop as any
}

function boolInput(prop: unknown, defaultVal = 1): number | (() => number) {
  if (prop === undefined) return defaultVal
  if (typeof prop === 'boolean') return prop ? 1 : 0
  if (typeof prop === 'function') return () => (prop as () => boolean)() ? 1 : 0
  if (isReactive(prop)) return () => unwrap(prop as any) ? 1 : 0
  return prop ? 1 : 0
}

// =============================================================================
// ENUM CONVERSIONS
// =============================================================================

function alignSelfToNum(a: string | undefined): number {
  switch (a) {
    case 'auto': return 0
    case 'flex-start': return 1
    case 'flex-end': return 2
    case 'center': return 3
    case 'baseline': return 4
    case 'stretch': return 5
    default: return 0 // auto
  }
}

function textAlignToNum(align: string | undefined): number {
  switch (align) {
    case 'center': return 1
    case 'right': return 2
    default: return 0 // left
  }
}

function textWrapToNum(wrap: string | undefined): number {
  switch (wrap) {
    case 'nowrap': return 0
    case 'truncate': return 2
    default: return 1 // wrap
  }
}

function justifySelfToNum(j: string | undefined): number {
  switch (j) {
    case 'start': return 1
    case 'end': return 2
    case 'center': return 3
    case 'stretch': return 4
    default: return 0 // auto
  }
}

/** Parse grid line position to i16 value */
function parseGridLine(line: GridLine | undefined): number {
  if (line === undefined || line === 'auto') return 0
  if (typeof line === 'number') return line
  if (typeof line === 'string' && line.startsWith('span ')) {
    return -parseInt(line.slice(5), 10) // negative = span
  }
  return 0
}

// =============================================================================
// TEXT POOL WRITER
// =============================================================================

/**
 * Write text to the text pool via setText() helper.
 * Returns the text offset for the repeater (reads from node after write).
 */
function writeTextToPool(buf: SharedBuffer, index: number, text: string): number {
  const result = setText(buf, index, text)

  if (!result.success) {
    const { liveBytes, poolSize, needed } = result
    const liveMB = (liveBytes / 1024 / 1024).toFixed(2)
    const poolMB = (poolSize / 1024 / 1024).toFixed(2)
    throw new Error(
      `Text pool full (${liveMB}MB live / ${poolMB}MB total). ` +
      `Cannot allocate ${needed} bytes for node ${index}. ` +
      `Increase textPoolSize in mount() config.`
    )
  }

  // Read the actual offset from the node (may differ due to slot reuse or compaction)
  return getU32(buf, index, N_TEXT_OFFSET)
}

// =============================================================================
// TEXT COMPONENT
// =============================================================================

export function text(props: TextProps): Cleanup {
  const buf = getBuffer()
  const arrays = getArrays()
  const index = allocateIndex(props.id)
  const disposals: (() => void)[] = []
  const parentIdx = getCurrentParentIndex()

  pushCurrentComponent(index)

  // --------------------------------------------------------------------------
  // CORE
  // --------------------------------------------------------------------------
  arrays.componentType.set(index, ComponentType.TEXT)

  // Set parent index and register in O(1) linked list
  arrays.parentIndex.set(index, parentIdx)
  registerParent(index, parentIdx)

  // Visibility (default: visible)
  disposals.push(repeat(boolInput(props.visible, 1), arrays.visible, index))

  // --------------------------------------------------------------------------
  // TEXT CONTENT — single repeater, no effects
  // --------------------------------------------------------------------------
  if (isReactive(props.content)) {
    disposals.push(repeat(
      () => writeTextToPool(buf, index, String(unwrap(props.content))),
      arrays.textOffset,
      index
    ))
  } else {
    // Static text — write once, no repeater needed
    const result = setText(buf, index, String(props.content))
    if (!result.success) {
      const { liveBytes, poolSize, needed } = result
      const liveMB = (liveBytes / 1024 / 1024).toFixed(2)
      const poolMB = (poolSize / 1024 / 1024).toFixed(2)
      throw new Error(
        `Text pool full (${liveMB}MB live / ${poolMB}MB total). ` +
        `Cannot allocate ${needed} bytes for node ${index}. ` +
        `Increase textPoolSize in mount() config.`
      )
    }
  }

  // --------------------------------------------------------------------------
  // LAYOUT — dimensions, flex item
  // --------------------------------------------------------------------------
  if (props.width !== undefined) disposals.push(repeat(dimInput(props.width), arrays.width, index))
  if (props.height !== undefined) disposals.push(repeat(dimInput(props.height), arrays.height, index))
  if (props.minWidth !== undefined) disposals.push(repeat(dimInput(props.minWidth), arrays.minWidth, index))
  if (props.maxWidth !== undefined) disposals.push(repeat(dimInput(props.maxWidth), arrays.maxWidth, index))
  if (props.minHeight !== undefined) disposals.push(repeat(dimInput(props.minHeight), arrays.minHeight, index))
  if (props.maxHeight !== undefined) disposals.push(repeat(dimInput(props.maxHeight), arrays.maxHeight, index))

  // Flex item
  if (props.grow !== undefined) disposals.push(repeat(numInput(props.grow), arrays.flexGrow, index))
  if (props.shrink !== undefined) disposals.push(repeat(numInput(props.shrink), arrays.flexShrink, index))
  if (props.flexBasis !== undefined) disposals.push(repeat(dimInput(props.flexBasis), arrays.flexBasis, index))
  if (props.alignSelf !== undefined) disposals.push(repeat(enumInput(props.alignSelf, alignSelfToNum), arrays.alignSelf, index))

  // Padding
  if (props.padding !== undefined) {
    disposals.push(repeat(numInput(props.paddingTop ?? props.padding), arrays.paddingTop, index))
    disposals.push(repeat(numInput(props.paddingRight ?? props.padding), arrays.paddingRight, index))
    disposals.push(repeat(numInput(props.paddingBottom ?? props.padding), arrays.paddingBottom, index))
    disposals.push(repeat(numInput(props.paddingLeft ?? props.padding), arrays.paddingLeft, index))
  } else {
    if (props.paddingTop !== undefined) disposals.push(repeat(numInput(props.paddingTop), arrays.paddingTop, index))
    if (props.paddingRight !== undefined) disposals.push(repeat(numInput(props.paddingRight), arrays.paddingRight, index))
    if (props.paddingBottom !== undefined) disposals.push(repeat(numInput(props.paddingBottom), arrays.paddingBottom, index))
    if (props.paddingLeft !== undefined) disposals.push(repeat(numInput(props.paddingLeft), arrays.paddingLeft, index))
  }

  // Margin
  if (props.margin !== undefined) {
    disposals.push(repeat(numInput(props.marginTop ?? props.margin), arrays.marginTop, index))
    disposals.push(repeat(numInput(props.marginRight ?? props.margin), arrays.marginRight, index))
    disposals.push(repeat(numInput(props.marginBottom ?? props.margin), arrays.marginBottom, index))
    disposals.push(repeat(numInput(props.marginLeft ?? props.margin), arrays.marginLeft, index))
  } else {
    if (props.marginTop !== undefined) disposals.push(repeat(numInput(props.marginTop), arrays.marginTop, index))
    if (props.marginRight !== undefined) disposals.push(repeat(numInput(props.marginRight), arrays.marginRight, index))
    if (props.marginBottom !== undefined) disposals.push(repeat(numInput(props.marginBottom), arrays.marginBottom, index))
    if (props.marginLeft !== undefined) disposals.push(repeat(numInput(props.marginLeft), arrays.marginLeft, index))
  }

  // Gap
  if (props.gap !== undefined) disposals.push(repeat(numInput(props.gap), arrays.gap, index))
  if (props.rowGap !== undefined) disposals.push(repeat(numInput(props.rowGap), arrays.rowGap, index))
  if (props.columnGap !== undefined) disposals.push(repeat(numInput(props.columnGap), arrays.columnGap, index))

  // Z-index
  if (props.zIndex !== undefined) disposals.push(repeat(numInput(props.zIndex), arrays.zIndex, index))

  // Text styling
  if (props.align !== undefined) disposals.push(repeat(enumInput(props.align, textAlignToNum), arrays.textAlign, index))
  if (props.wrap !== undefined) disposals.push(repeat(enumInput(props.wrap, textWrapToNum), arrays.textWrap, index))

  // --------------------------------------------------------------------------
  // GRID ITEM PROPERTIES
  // --------------------------------------------------------------------------
  if (props.gridColumnStart !== undefined) {
    if (isReactive(props.gridColumnStart)) {
      disposals.push(repeat(() => parseGridLine(unwrap(props.gridColumnStart)), arrays.gridColumnStart, index))
    } else {
      arrays.gridColumnStart.set(index, parseGridLine(props.gridColumnStart as GridLine))
    }
  }
  if (props.gridColumnEnd !== undefined) {
    if (isReactive(props.gridColumnEnd)) {
      disposals.push(repeat(() => parseGridLine(unwrap(props.gridColumnEnd)), arrays.gridColumnEnd, index))
    } else {
      arrays.gridColumnEnd.set(index, parseGridLine(props.gridColumnEnd as GridLine))
    }
  }
  if (props.gridRowStart !== undefined) {
    if (isReactive(props.gridRowStart)) {
      disposals.push(repeat(() => parseGridLine(unwrap(props.gridRowStart)), arrays.gridRowStart, index))
    } else {
      arrays.gridRowStart.set(index, parseGridLine(props.gridRowStart as GridLine))
    }
  }
  if (props.gridRowEnd !== undefined) {
    if (isReactive(props.gridRowEnd)) {
      disposals.push(repeat(() => parseGridLine(unwrap(props.gridRowEnd)), arrays.gridRowEnd, index))
    } else {
      arrays.gridRowEnd.set(index, parseGridLine(props.gridRowEnd as GridLine))
    }
  }
  if (props.justifySelf !== undefined) {
    disposals.push(repeat(enumInput(props.justifySelf, justifySelfToNum), arrays.justifySelf, index))
  }

  // --------------------------------------------------------------------------
  // VISUAL — colors with variant support
  // --------------------------------------------------------------------------
  if (props.variant && props.variant !== 'default') {
    const variant = props.variant
    disposals.push(repeat(
      props.fg !== undefined ? colorInput(props.fg) : () => toPackedColor(getVariantStyle(variant).fg),
      arrays.fgColor, index
    ))
    disposals.push(repeat(
      props.bg !== undefined ? colorInput(props.bg) : () => toPackedColor(getVariantStyle(variant).bg),
      arrays.bgColor, index
    ))
  } else {
    if (props.fg !== undefined) disposals.push(repeat(colorInput(props.fg), arrays.fgColor, index))
    if (props.bg !== undefined) disposals.push(repeat(colorInput(props.bg), arrays.bgColor, index))
  }
  if (props.opacity !== undefined) disposals.push(repeat(numInput(props.opacity), arrays.opacity, index))

  // --------------------------------------------------------------------------
  // MOUSE HANDLERS
  // --------------------------------------------------------------------------
  let unsubMouse: (() => void) | undefined

  if (props.onMouseDown || props.onMouseUp || props.onClick || props.onMouseEnter || props.onMouseLeave || props.onScroll) {
    unsubMouse = onMouseComponent(index, {
      onMouseDown: props.onMouseDown,
      onMouseUp: props.onMouseUp,
      onClick: props.onClick,
      onMouseEnter: props.onMouseEnter,
      onMouseLeave: props.onMouseLeave,
      onScroll: props.onScroll,
    })
  }

  // Component setup complete
  popCurrentComponent()
  runMountCallbacks(index)

  // --------------------------------------------------------------------------
  // CLEANUP
  // --------------------------------------------------------------------------
  const cleanup = () => {
    for (const dispose of disposals) dispose()
    disposals.length = 0
    unsubMouse?.()
    cleanupKeyboardListeners(index)
    releaseIndex(index)
  }

  const scope = getActiveScope()
  if (scope) scope.cleanups.push(cleanup)

  return cleanup
}
